sig-api-machinery/extensible-admission-to-beta.md

```diff
@@ -0,0 +1,363 @@
+  based on either a namespace-level annotation, or the plugin's configured value
+  (although this probably should be part of LimitRange).
+- apply quota at the cluster level instead of per-namespace
+- assign default resource request values relative to user-specified limits for
+  pods to support an administrator-defined overcommit target
+- restrict who is allowed to set the `nodeName` or `nodeSelector` on pods
+- ensure you can't exec or attach to a pod with a privileged security context if
+  you don't have privileges to create a privileged pod
+- supply default values for certain fields on Build Pods when unset (e.g.
+  environment variables)
+- enforce cluster-wide configuration values on all Build Pods (i.e.
+  override user-supplied data if necessary)
+- automatically provision a Jenkins server in a user's namespace if it doesn't
+  exist when a pipeline BuildConfig is created
+
+We would like to have the ability to run a single external plugin at the end of
+each admission phase. This will enable cluster administrators to inject custom
+admission logic into an existing kube-apiserver without having to recompile it.
+
+### OPEN QUESTIONS
+
+#### What happens if the external admission plugin is not available?
+
+Let's imagine that I have a Kubernetes cluster up and running. Now I want to run
+OpenShift on top of it, so I deploy its various pods to the cluster, presumably
+as Deployment resources. Next, I need to add OpenShift as an external admission
+plugin for the kube-apiserver pods and restart them.
+
+Fast forward a bit - now I have everything working. Then let's imagine that my
+openshift-apiserver pods fail for whatever reason. What happens to
+kube-apiserver requests with respect to admission? With openshift-apiserver
+down, any attempt by the kube-apiserver to invoke the external admission plugin
+will fail.
+
+Do we ignore an external admission plugin being unreachable and allow requests
+to proceed ("fail open")? I don't think this is viable. Presumably the external
+admission plugin is there for a reason, and if it is unable to execute, it opens
+the possibility for inconsistent or incorrect data to enter the cluster.
+
+Do we instead reject the request because the external admission plugin is
+unreachable ("fail closed")? I think this is what we must do, for the reason
+stated above. Assuming this is what we choose, how do we restore the cluster to
+a healthy state, as the default "fail closed" behavior won't allow any new pods,
+and we need at least one new openshift-apiserver pod to handle the external
+admission plugin invocations. We need some way to a) indicate that a certain set
+of changes to the system are being made to restore the external admission
+plugin, and b) ensure that users can't "sneak in" anything while admission isn't
+fully operational. Some possible options include:
+
+- Specify via flag/configuration the identity of a user or group whose requests
```

Comment:
What if we register the namespace that the admission plugin lives in, and say that any given plugin does not apply to objects in its own namespace? Then you use regular RBAC, e.g., to control who has access to create things in that namespace.